doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
+
+   Internally "x <> y" is represented as "(~ (x = y))", which produces the
+   first match.
+
+   .. coqtop:: in
+
+      Ltac2 f2 t := match! t with
+                    | context [ (~ ?t) ] => print (of_constr t); fail
+                    | _ => ()
+                    end.
+
+   .. coqtop:: all abort
+
+      f2 constr:((~ True) <> (~ False)).
+
 Match over goals
 ~~~~~~~~~~~~~~~~
 
-Similarly, there is a way to match over goals in an elegant way, which is
-just a notation desugared at parsing time.
+.. tacn:: @ltac2_match_key {? reverse } goal with @goal_match_list end
+   :name: lazy_match! goal; match! goal; multi_match! goal
 
-.. productionlist:: coq
-   ltac2_term             : match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
-                    : lazy_match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
-                    : multi_match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
-   goalmatching     : | [ `hypmatching` ... `hypmatching` |- `constrpattern` ] => `ltac2_term`
-   hypmatching      : `lident` : `constrpattern`
-                    : _ : `constrpattern`
+   .. insertprodn goal_match_list gmatch_hyp_pattern
 
-Variables from :n:`@hypmatching` and :n:`@constrpattern` are bound in the body of the
-branch. Their types are:
+   .. prodn::
+      goal_match_list ::= {? %| } {+| @gmatch_rule }
+      gmatch_rule ::= @gmatch_pattern => @ltac2_expr
+      gmatch_pattern ::= [ {*, @gmatch_hyp_pattern } |- @ltac2_match_pattern ]
+      gmatch_hyp_pattern ::= @name : @ltac2_match_pattern
 
-- ``constr`` for pattern variables appearing in a :n:`@term`
-- ``Pattern.context`` for variables binding a context
-- ``ident`` for variables binding a hypothesis name.
+   Matches over goals, similar to Ltac1 :tacn:`match goal`.
+   Use this form to match hypotheses and/or goals in the proof context.  These patterns have zero or
+   more subpatterns to match hypotheses followed by a subpattern to match the conclusion.  Except for the
+   differences noted below, this works the same as the corresponding :n:`@ltac2_match_key @ltac2_expr` construct
+   (see :tacn:`match!`).  Each current goal is processed independently.
 
-The same identifier caveat as in the case of matching over constr applies, and
-this features has the same semantics as in Ltac1. In particular, a ``reverse``
-flag can be specified to match hypotheses from the more recently introduced to
-the least recently introduced one.
+   :n:`{*, @gmatch_pattern }`
+      Patterns to match with hypotheses.  Each pattern must match a distinct hypothesis in order
+      for the branch to match.
 
-.. _ltac2_notations:
+      Hypotheses have the form :n:`@name {? := @term__binder } : @type`.  Currently Ltac2 doesn't
+      allow matching on or capturing the value of :n:`@term__binder`.  It only supports matching on
+      the :token:`name` and the :token:`type`, for example `n : ?t`.
```

Comment:
This seems like a rather severe restriction.  Is there an issue open for it?

Reply:
I didn't file one or check whether there already is one.  Feel free to file one.

Reply:
Reported as https://github.com/coq/coq/issues/12803